modules/sitemaps/sitemaps.php

```diff
@@ -1,782 +1,1556 @@
+		$log = new Jetpack_Sitemap_Logger('begin news sitemap generation');
+
+		$buffer = new Jetpack_Sitemap_Buffer(
+			self::NEWS_SITEMAP_MAX_ITEMS,
+			self::SITEMAP_MAX_BYTES,
+
+			/* open tag */
+			"<?xml version='1.0' encoding='UTF-8'?>\n" .
+			"<!-- generator='jetpack-" . JETPACK__VERSION . "' -->\n" .
+			"<?xml-stylesheet type='text/xsl' href='" . home_url() . "/news-sitemap.xsl'?>\n" .
+			"<urlset\n" .
+			"  xmlns='http://www.sitemaps.org/schemas/sitemap/0.9'\n" .
+			"  xmlns:news='http://www.google.com/schemas/sitemap-news/0.9'>\n",
+
+			/* close tag */
+			"</urlset>\n",
+
+			/* epoch */
+			strtotime('1970-01-01 00:00:00')
+		);
+
+		// Retrieve the 1000 most recent posts.
+		$posts = $this->query_most_recent_posts(1000);
+
+		// For each post in the batch,
+		foreach ($posts as $post) {
+			// Generate the sitemap XML for the post.
+			$current_item = $this->post_to_news_sitemap_item($post);
+
+			// Try to add it to the buffer.
+			if ( false == $buffer->try_to_add_item($current_item['xml']) ) {
+				break;
+			}
+		}
+
+		$log->time('end news sitemap generation');
+
+		return $buffer->contents();
+	}
+
+	/**
+	 * Construct the sitemap url entry for a WP_Post.
 	 *
-	 * @since 3.9.0
+	 * @link http://www.sitemaps.org/protocol.html#urldef
+	 * @access private
+	 * @since 4.5.0
 	 *
-	 * @param array $blog_home Data to build parent and children nodes for site home.
+	 * @param WP_Post $post The post to be processed.
+	 * @return string An XML fragment representing the post URL.
 	 */
-	$url_node = apply_filters( 'jetpack_sitemap_url_home', $blog_home );
-	jetpack_sitemap_array_to_simplexml( array( 'url' => $url_node ), $tree );
-	unset( $blog_home );
+	private function post_to_sitemap_item ( $post ) {
+		$url = get_permalink($post);
+
+		/*
+		 * Must use W3C Datetime format per the spec.
+		 * https://www.w3.org/TR/NOTE-datetime
+		 */ 
+		$last_modified = str_replace( ' ', 'T', $post->post_date) . 'Z';
+
+		/*
+		 * Spec requires the URL to be <=2048 bytes.
+		 * In practice this constraint is unlikely to be violated.
+		 */
+		if ( mb_strlen($url) > 2048 ) {
+			$url = site_url() . '/?p=' . $post->ID; 
+		}
+
+		$xml =
+			"<url>\n" .
+			" <loc>$url</loc>\n" .
+			" <lastmod>$last_modified</lastmod>\n" .
+			"</url>\n";
+
+		return array(
+			'xml'           => $xml,
+			'last_modified' => $post->post_date
+		);
+	}
 
 	/**
-	 * Filter data before rendering it as XML.
+	 * Construct the image sitemap url entry for a WP_Post of image type.
+	 *
+	 * @link http://www.sitemaps.org/protocol.html#urldef
 	 *
-	 * @module sitemaps
+	 * @access private
+	 * @since 4.5.0
 	 *
-	 * @since 3.9.0
+	 * @param WP_Post $post The image post to be processed.
 	 *
-	 * @param SimpleXMLElement $tree Data tree for sitemap.
-	 * @param string $latest_mod Date of last modification.
+	 * @return string An XML fragment representing the post URL.
 	 */
-	$tree = apply_filters( 'jetpack_print_sitemap', $tree, $latest_mod );
```

Comment:
Some site owners already use that filter, so I think it would be nice to keep it.